Documentation/network/lb-ipam.rst

```diff
@@ -50,11 +55,26 @@ After adding the pool to the cluster, it appears like so.
     NAME        DISABLED   CONFLICTING   IPS AVAILABLE   AGE
     blue-pool   false      False         65788           2s
 
-.. note::
-    The amount of available IPs in the pool is lower than the actual sum of all
-    usable IPs in the CIDRs because the allocation logic is limited to 65536 IPs
-    per CIDR. CIDRs containing more than 65536 IPs can be broken down into multiple
-    smaller CIDRs to achieve full utilization.
+CIDRs, Ranges and dodgy IPs
+---------------------------
+
+An IP pool can have multiple blocks of IPs. A block can be specified with CIDR
+notation (<prefix>/<bits>) or a range notation with a start and stop IP. As
+pictured in :ref:`lb_ipam_pools`.
+
+CIDRs are often used to specify routable IP ranges. By convention, the first
+and the last IP of a CIDR are reserved. The first IP is the 
+"network address" and the last IP is the "broadcast address". In most networks
+these IPs are not usable. In LB-IPAM we refer to these as "Dodgy IPs" because
+they do not always play well with all network equipment. LB-IPAM will not assign
+these by default. Exceptions are /32 and /31 IPv4 CIDRs and /128 and /127 IPv6
+CIDRs since these only have 1 or 2 IPs respectively.
+
+If you wish to use these dodgy IPs, you can set the ``.spec.allowDodgyIPs`` field
```

Comment:
If we configure an expensive address pool such as global ipv4 addresses, we must set `allowDodgyIPs=true`, correct? If so, I'm not sure this spec makes sense because there are more cases where you want to set `allowDodgyIPs=false` from my experience. Plus, I'm confused because it's the opposite of the MetalLB `AvoidBuggyIPs` flag.

Reply:
> If we configure an expensive address pool such as global ipv4 addresses, we must set allowDodgyIPs=true, correct? If so, I'm not sure this spec makes sense because there are more cases where you want to set allowDodgyIPs=false from my experience.

Yea, that is correct, if you want to use the first and last IP of the global range then you have to set `allowDodgyIPs` to `true`. If unspecified, the value defaults to `false`, so unless you specify `allowDodgyIPs=true`, the first and last IPs are reserved, the same behavior we currently have.

> I'm confused because it's the opposite of the MetalLB `AvoidBuggyIPs` flag.

Yea, this can't really be helped because we need to maintain backwards compatibility for existing policies, otherwise we may break setups on upgrade.

> We use /28 for global and bastion networks. So I assume we need to set allowDodgyIPs=true accordingly.

Practical example for `10.0.0.128/28`, if set to `true`, the allocate range is `10.0.0.128-10.0.0.143`, if `false` or unspecified its `10.0.0.129-10.0.0.142`. So you should pick which is desirable for your scenario

Reply:
Thank you for your comment!

I'm worried that users migrating from MetalLB(either MetalLB BGP ControlPlane or standalone) will be confused. With that said, it is understandable that backward compatibility must be maintained.

Wouldn't it be acceptable to break compatibility after reminding users in the upgrade guide in this case?

Reply:
> Wouldn't it be acceptable to break compatibility after reminding users in the upgrade guide in this case?

It could be in some cases, but consider the worst scenario for this case if a user doesn't pay attention or misses the notice. Any user with existing pools will suddenly be able to allocate the dodgy IPs, hard to notice if you are not looking for it. Lets say a service gets assigned a broadcast IP in a network which has this feature enabled. At best the feature doesn't work, at worst it causes a DoS or security issue.

I think the confusion concern is valid, but I hope it isn't as big a deal as we might think. I would like to know what others think.

Reply:
I think I'm slightly biased because I'm the user planning to migrate from MetalLB. So yeah, we should ask what others think.

Reply:
I would change this behaviour and enable this by default. Here's my rationale:
network and broadcast addresses only make sense on a network segment. When we allocate and announce individual host prefixes, they have no notion of broadcast or network addresses. 
For example, 192.168.255.255 may be a broadcast address for a segment with prefix 192.168.0.0/16 but may be a normal address for prefix 192.168.0.0/15. 
Even if we announce individual host IPs and they get aggregated on a border router, they still are not assigned to any network segment (aggregate route is configured as null-route).

The only caveat may be with L2 announcer where, if the range is picked incorrectly, allocated IP can overlap with local subnet's broadcast address. While this is more of an operational problem, I don't think this will result in any loss of connectivity too.
---
I've got an extra ask, can we please not use the word `dodgy`? This is not a generally-used term and may get a lot of people confused. I can suggest `allowFirstLastlIPs` or `prefixAsRange` or `CIDRasRange`.

Reply:
I brought this issue up in last weeks Community meeting to ask for advice regarding breakage of backwards compatibility. I think we settled on a decent middle ground, that is, in 1.15 we will not break anything, introduce a setting to change the behavior. Then next release, 1.16 we will switch over the default.

We will add warnings in the feature docs and upgrade guide, warning people of the upcoming change, and asking people to explicitly set what they want. That way, we have an upgrade path where we don't break setups on upgrade and change the behavior even if it takes a bit.

I am thinking of changing the setting from a boolean to a enum, so we get something like `allowFirstLastIPs: yes|no|default` with `default` being the value chosen if the field is not specified. This also should make the field different enough from MetalLB's implementation to not cause confusion.

Would this be acceptable?

Reply:
@dylandreimerink Thank you for bringing this up!

So, in 1.15, we will introduce `allowFirstLastIPs=false` by default to ensure backward compatibility, and in 1.16 we will change its default, which is  `allowFirstLastIPs=true` by default. Is my understanding correct?

Reply:
Yes. Currently there is no setting, so if we add the setting and change the default, then users have no chance to set the setting before the default switches. So by doing it in two versions, we allow users some time to explicitly set what they want the behavior to be for existing pools before we switch.

Reply:
Alright, should be good now